### DISCUSSION TITLE
Playtime/ReaLearn trigger at start/end of a clip/measure

### DISCUSSION THREAD

**Discussion by Whiskeysaurus** [helgoboss/helgobox#1207] [General]
In Playtime, Say, for example, you wanted to clear a slot/delete a clip only when it ends. A broader example, say you wanted to automate arming column 2 and recording a new clip so that it happens when a clip on column 1 starts playing. Is there a way to do this?

A simpler version perhaps, is there a way to quantize a command to run at the beginning of a Playtime measure?

**Comment by helgoboss** [comment]
Don't think so. What's your musical goal here? Maybe there's another way.

**Comment by Whiskeysaurus** [comment]
Live looping performance in Playtime. Basically just need to cue up actions to happen at the end of a measure so that they're perfectly in time. I figure that Playtime clips would be the best way to do this, but maybe there's a workaround? I know Follow Actions are planned (#1044.) Is there a way that a custom OSC message could be added as a follow action? That seems like a quick way to integrate them into ReaLearn.

**Comment by helgoboss** [comment]
The problem is that generally executing some action to happen at the end of some event, e.g. after clip stop, will not necessarily have good timing (due to dealing with different processing threads).

So the best thing is if you tell me what concrete scenarios you would like implement with such a possibility. Walk me through all scenarios that you are looking to achieve. Which actions do you plan to cue up and to what end?

**Comment by Whiskeysaurus** [comment]
An ideal example would be automatic cues for a live looping performance like is possible in Ableton. 
IE record track A for 2 measures, then automatically arm track B and record for 2 bars, and so on. All without having to use a foot pedal or something.

A more specific important case for my use in switching from Ableton is queuing the deletion and recording of a slot, for replacing the current loop on the fly with no gap.

**Reply by helgoboss** [reply]
For now, there's no supported way to achieve any of that (maybe some workarounds, but which will probably be non-ideal in terms of timing).

But yes, follow actions will come at some point. So the question is: Which of those use cases would be covered by follow actions? And which wouldn't?

**Reply by Whiskeysaurus** [reply]
Depends on what the available follow actions would be. I'd say in the short term, you could get near-infinite mileage out of just a single follow action that triggers a list of given ReaLearn mappings ordered and selected via their IDs. Then it's perhaps a bit tedious for the user, but they can build their own follow actions however they need. Same could be accomplished via a list of custom OSC messages too.

**Reply by helgoboss** [reply]
That's an interesting suggestion and I can understand that this would be desirable. However, the timing issue still stands. So OSC would be a no go. And direct triggering of a ReaLearn mapping would only work if the target in question supports real-time control. And even then ... it would be only precise on block level, not on sample level. Which is not optimal or even unacceptable for many scenarios.

I suggest you go over to the follow-actions FR and add a comment there with a list of all conrete actions that you would like to see. When I implement follow actions at some point in the future, I will see that list and try to consider it.
